metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -104,53 +117,73 @@ public PartitionChangeBuilder setTargetAdding(List<Integer> targetAdding) {
-        // If the new isr doesn't have the current leader, we need to try to elect a new
-        // one. Note: this also handles the case where the current leader is NO_LEADER,
-        // since that value cannot appear in targetIsr.
-        if (!targetIsr.contains(partition.leader)) return true;
-
-        // Check if we want to try to get away from a non-preferred leader.
-        if (alwaysElectPreferredIfPossible && !partition.hasPreferredLeader()) return true;
+    // VisibleForTesting
+    static class ElectionResult {
+        final int node;
+        final boolean unclean;
 
-        return false;
+        private ElectionResult(int node, boolean unclean) {
+            this.node = node;
+            this.unclean = unclean;
+        }
     }
 
-    class BestLeader {
-        final int node;
-        final boolean unclean;
+    // VisibleForTesting
+    /**
+     * Perform leader election based on the partition state and leader election type.
+     *
+     * See documentation for the Election type to see more details on the election types supported.
+     */
+    ElectionResult electLeader() {
+        // 1. Check if the election is not PREFERRED and we already have a valid leader
+        if (election != Election.PREFERRED && targetIsr.contains(partition.leader) && isAcceptableLeader.apply(partition.leader)) {
+            // Don't consider a new leader since the current leader meets all the constraints
+            return new ElectionResult(partition.leader, false);
+        }
 
-        BestLeader() {
-            for (int replica : targetReplicas) {
-                if (targetIsr.contains(replica) && isAcceptableLeader.apply(replica)) {
-                    this.node = replica;
-                    this.unclean = false;
-                    return;
-                }
-            }
-            if (uncleanElectionOk.get()) {
-                for (int replica : targetReplicas) {
-                    if (isAcceptableLeader.apply(replica)) {
-                        this.node = replica;
-                        this.unclean = true;
-                        return;
-                    }
-                }
+        // 2. Attempt preferred replica election
+        int preferredReplica = targetReplicas.get(0);
+        if (targetIsr.contains(preferredReplica) && isAcceptableLeader.apply(preferredReplica)) {
+            return new ElectionResult(preferredReplica, false);
+        }
+
+        // 3. Preferred replica was not elected, only continue if the current leader is not a valid leader
+        if (targetIsr.contains(partition.leader) && isAcceptableLeader.apply(partition.leader)) {
+            // Don't consider a new leader since the current leader meets all the constraints
+            return new ElectionResult(partition.leader, false);
+        }
+
+        // 4. Attempt to keep the partition online based on the ISR
+        Optional<Integer> bestLeader = targetReplicas.stream()
+            .skip(1)
```

Comment:
Hmm, why do we want to skip the first replica? When election != Election.PREFERRED, we still want to consider the first replica as long as it's in ISR and is alive.

Reply:
We can skip the first replica because we already considered it in 2. or lines 144-148. This is a small optimization.